### PR TITLE
Add an optional filter to `snapshot` for frames to include and fix unserializable values

### PR DESCRIPTION
**Pull request by leowrites**
## Proposed Changes

- Add an optional filter `include` to `snapshot` to only capture variables in specified frames
- Use `repr` to serialize certain unserializable values (e.g. `PosixPath`)

...

## Type of Change

_(Write an `X` or a brief description next to the type or types that best describe your changes.)_

| Type                                                                                    | Applies? |
| --------------------------------------------------------------------------------------- | -------- |
| 🚨 _Breaking change_ (fix or feature that would cause existing functionality to change) |          |
| ✨ _New feature_ (non-breaking change that adds functionality)                          |      x    |
| 🐛 _Bug fix_ (non-breaking change that fixes an issue)                                  |    x      |
| ♻️ _Refactoring_ (internal change to codebase, without changing functionality)          |          |
| 🚦 _Test update_ (change that _only_ adds or modifies tests)                            |      x    |
| 📚 _Documentation update_ (change that _only_ updates documentation)                    |      x    |
| 📦 _Dependency update_ (change that updates a dependency)                               |          |
| 🔧 _Internal_ (change that _only_ affects developers or continuous integration)         |          |

## Checklist

_(Complete each of the following items for your pull request. Indicate that you have completed an item by changing the `[ ]` into a `[x]` in the raw text, or by clicking on the checkbox in the rendered description on GitHub.)_

Before opening your pull request:

- [x] I have performed a self-review of my changes.
  - Check that all changed files included in this pull request are intentional changes.
  - Check that all changes are relevant to the purpose of this pull request, as described above.
- [x] I have added tests for my changes, if applicable.
  - This is **required** for all bug fixes and new features.
- [x] I have updated the project documentation, if applicable.
  - This is **required** for new features.
- [x] I have updated the project Changelog (this is required for all changes).
- [ ] If this is my first contribution, I have added myself to the list of contributors.

After opening your pull request:

- [x] I have verified that the pre-commit.ci checks have passed.
- [ ] I have verified that the CI tests have passed.
- [ ] I have reviewed the test coverage changes reported by Coveralls.
- [x] I have [requested a review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) from a project maintainer.

## Questions and Comments

_(Include any questions or comments you have regarding your changes.)_

### CODE REVIEW

File: python_ta/debug/snapshot.py [hunk 1]
```diff
@@ -40,6 +41,7 @@ def snapshot(
     save: bool = False,
     memory_viz_args: Optional[list[str]] = None,
     memory_viz_version: str = "latest",
+    include: Optional[list[str]] = [".*"],
```
Comment (leowrites) [inline python_ta/debug/snapshot.py:44]: Using a mutable default value is not the best choice, see if it's possible to change it to a tuple

File: tests/test_debug/test_snapshot.py [hunk 2]
```diff
@@ -642,3 +644,82 @@ def test_snapshot_save_stdout():
...
+
+def func_with_unserializable_object():
+    path = pathlib.PosixPath("some path")
+    vars_in_func = [snapshot()[0]]
+    processed_result = snapshot_to_json(vars_in_func)
+    json.dumps(processed_result)
+    return processed_result
+
+
+def test_snapshot_only_include_called_function():
```
Comment (leowrites) [inline tests/test_debug/test_snapshot.py:669]: should be "includes"

File: tests/test_debug/test_snapshot.py [hunk 3]
```diff
@@ -642,3 +644,82 @@ def test_snapshot_save_stdout():
...
+            "func_with_include_nested": {
+                "include": ["func_with_include_nested"],
+                "test_var1b": {"SDS_coolest_project": "PyTA"},
+                "test_var2b": ("Leo", "tester"),
+            },
+        }
+    ]
+
+
+def test_snapshot_serializes_unserializable_value():
```
Comment (leowrites) [inline tests/test_debug/test_snapshot.py:715]: "unserializable values"

File: tests/test_debug/test_snapshot.py [hunk 4]
```diff
@@ -642,3 +644,93 @@ def test_snapshot_save_stdout():
         expected_svg = expected_svg_file.read()
 
     assert result.stdout == expected_svg
+
+
+def func_with_include(include: Optional[tuple[str, ...]] = None):
```
Comment (david-yz-liu) [inline tests/test_debug/test_snapshot.py:649]: See above note about type annotation of `include` parameter. Also add a return type annotation.

(Applies to all helper functions defined here.)

File: python_ta/debug/snapshot.py [hunk 5]
```diff
@@ -40,6 +41,7 @@ def snapshot(
     save: bool = False,
     memory_viz_args: Optional[list[str]] = None,
     memory_viz_version: str = "latest",
+    include: Optional[tuple[str, ...]] = (".*",),
```
Comment (david-yz-liu) [inline python_ta/debug/snapshot.py:44]: Make the default `None`, and in the code below if `include is None` then skip the checks.

Make the inner type annotation an `Iterable[str]` (import from `typing`) to allow different kinds of collections to be passed in.

Also, just to be complete, the contained type can be `str | re.Pattern`, as either of those types are permitted by `re.compile`.

File: python_ta/debug/snapshot.py [hunk 6]
```diff
@@ -51,16 +53,19 @@ def snapshot(
     For details on the MemoryViz CLI, see https://www.cs.toronto.edu/~david/memory-viz/docs/cli.
     memory_viz_version can be used to dictate version, with a default of the latest version.
     Note that this function is compatible only with MemoryViz version 0.3.1 and above.
+    include can be used to specify a list of function names which variables will be captured.
```
Comment (david-yz-liu) [inline python_ta/debug/snapshot.py:56]: 1. list -> collection
2. which -> whose
3. this doesn't mention regular expressions; so, more precise elaboration would be good

**Comment by david-yz-liu** [comment]
@leowrites one comment I forgot to mention: everywhere you're adding new types, you probably should include the `from __future__ import annotations` as the module's first line. This would fix the failing Python 3.8 CI tests, for example.

**Comment by leowrites** [comment]
@david-yz-liu Thank you for the feedback. I made the changes and `test_snapshot_to_json_sets_primitive` is failing which is a known issue.

**Review by david-yz-liu — APPROVED** [review]

**Pull request merged**